src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+using namespace web::websockets::client;
+using namespace concurrency::streams;
+
+namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
+{
+    AzureConnection::AzureConnection(const uint32_t initialRows, const uint32_t initialCols) :
+        _initialRows{ initialRows },
+        _initialCols{ initialCols },
+        _loginUri{ U("https://login.microsoftonline.com/") },
+        _resourceUri{ U("https://management.azure.com/") },
+        _wantedResource{ U("https://management.core.windows.net/") }
+    {
+    }
+
+    winrt::event_token AzureConnection::TerminalOutput(Microsoft::Terminal::TerminalConnection::TerminalOutputEventArgs const& handler)
+    {
+        return _outputHandlers.add(handler);
+    }
+
+    void AzureConnection::TerminalOutput(winrt::event_token const& token) noexcept
+    {
+        _outputHandlers.remove(token);
+    }
+
+    winrt::event_token AzureConnection::TerminalDisconnected(Microsoft::Terminal::TerminalConnection::TerminalDisconnectedEventArgs const& handler)
+    {
+        return _disconnectHandlers.add(handler);
+    }
+
+    void AzureConnection::TerminalDisconnected(winrt::event_token const& token) noexcept
+    {
+        _disconnectHandlers.remove(token);
+    }
+
+    void AzureConnection::Start()
+    {
+        // Create our own output handling thread
+        // Each connection needs to make sure to drain the output from its backing host.
+        _hOutputThread.reset(CreateThread(nullptr,
+                                          0,
+                                          StaticOutputThreadProc,
+                                          this,
+                                          0,
+                                          nullptr));
+
+        _connected = true;
+    }
+
+    void AzureConnection::WriteInput(hstring const& data)
+    {
+        if (!_connected || _closing.load())
+        {
+            return;
+        }
+
+        // This case only happens when the user has multiple tenants
+        // and we require their input for which tenant they want
+        if (_needInput)
+        {
+            int tenantNum = -1;
+            while (tenantNum < 0 || tenantNum >= _maxSize)
+            {
+                if (_closing.load())
+                {
+                    _needInput = false;
+                    return;
+                }
+                try
+                {
+                    tenantNum = std::stoi(winrt::to_string(data));
+                }
+                catch (...)
+                {
+                    _outputHandlers(winrt::to_hstring("Please enter a number.\r\n"));
+                    return;
+                }
+                if (tenantNum >= _maxSize)
+                {
+                    _outputHandlers(winrt::to_hstring("Number out of bounds.\r\n"));
+                    return;
+                }
+            }
+            std::lock_guard<std::mutex> lg{ _tenantNumMutex };
+            _tenantNumber = tenantNum;

```

Comment:
Consider using `std::atomic`

Reply:
We need a lock guard (and a condition variable) here because we want to hold the thread until we get a valid input from the user.